File: 4.5-ai-ivr/scripts/relay-server.php
<?php
// 3.4.5 AI IVR — Voice Relay server that classifies caller intent with an LLM
// and patches a human agent into the live call.
//
// Flow per connection:
//   1. Sinch sends {"command":"connect","callId":"..."} -> we answer + greet.
//   2. Sinch sends {"command":"text"/"prompt","text":"<caller speech>"}.
//   3. We ask the LLM to classify. One-word reply ("Sales"/"Support") = route;
//      anything longer is spoken back as a clarifying question.
//   4. On a route, we PATCH /v2/projects/{projectId}/calls/{callId} to dial the
//      agent and bridge them into "ivr-bridge", then close the socket.
//
// Requirements: composer install   (installs cboden/ratchet). PHP 8.1+ with curl.
// Run:    php relay-server.php
// Expose: ngrok http 8765

require __DIR__ . '/../vendor/autoload.php';

use Ratchet\ConnectionInterface;
use Ratchet\MessageComponentInterface;
use Ratchet\Server\IoServer;
use Ratchet\Http\HttpServer;
use Ratchet\WebSocket\WsServer;

// ── Load the tutorial-folder .env (../.env relative to this scripts/ folder) ──
$envPath = __DIR__ . '/../.env';
if (is_file($envPath)) {
    foreach (file($envPath, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES) as $line) {
        if ($line[0] === '#' || !str_contains($line, '=')) continue;
        [$k, $v] = explode('=', $line, 2);
        $k = trim($k);
        $v = trim(trim($v), "\"'");
        if (getenv($k) === false) putenv("$k=$v");
        $_ENV[$k] = $v;
    }
}

function env_require(string $name): string {
    $v = $_ENV[$name] ?? getenv($name);
    if ($v === false || $v === null || $v === '') {
        fwrite(STDERR, "ERROR: $name is not set in the environment / .env\n");
        exit(1);
    }
    return $v;
}

// ── Configuration ────────────────────────────────────────────────────────────
const SINCH_BASE = 'https://voice.api.sinch.com/v2';
const GREETING   = 'Hello, this is the call centre. How can I help you?';

$config = [
    'projectId'   => env_require('PROJECT_ID'),
    'keyId'       => env_require('KEY_ID'),
    'keySecret'   => env_require('KEY_SECRET'),
    'sinchNumber' => env_require('SINCH_NUMBER'),
    'llmBaseUrl'  => rtrim($_ENV['LLM_BASE_URL'] ?? getenv('LLM_BASE_URL') ?: 'https://api.openai.com/v1', '/'),
    'llmApiKey'   => env_require('LLM_API_KEY'),
    'llmModel'    => $_ENV['LLM_MODEL'] ?? getenv('LLM_MODEL') ?: 'gpt-4o-mini',
    'agents'      => [
        'sales'   => [env_require('SALES_NUMBER'),   'Sales'],
        'support' => [env_require('SUPPORT_NUMBER'), 'Support'],
    ],
    'systemPrompt' => trim(file_get_contents(__DIR__ . '/../system_prompt.md')),
];

$port = (int)($_ENV['PORT'] ?? getenv('PORT') ?: 8765);

// ── Relay component ──────────────────────────────────────────────────────────
class IvrRelay implements MessageComponentInterface
{
    private SplObjectStorage $state;     // conn => ['callId' => ?string, 'patched' => bool]
    private $loop = null;

    public function __construct(private array $cfg)
    {
        $this->state = new SplObjectStorage();
    }

    public function setLoop($loop): void { $this->loop = $loop; }

    public function onOpen(ConnectionInterface $conn): void
    {
        $this->state[$conn] = ['callId' => null, 'patched' => false];
        echo "[+] connected\n";
    }

    public function onMessage(ConnectionInterface $conn, $raw): void
    {
        echo "  << $raw\n";
        $msg = json_decode($raw, true);
        if (!is_array($msg)) return;
        $command = $msg['command'] ?? null;
        $st = $this->state[$conn];

        if ($command === 'connect') {
            $st['callId'] = $msg['callId'] ?? null;
            $this->state[$conn] = $st;
            $this->send($conn, ['command' => 'answer']);
            $this->send($conn, ['command' => 'text', 'text' => GREETING, 'isLast' => true]);
            return;
        }

        if (in_array($command, ['text', 'prompt'], true) && $st['callId'] && !$st['patched']) {
            $callerText = trim($msg['text'] ?? '');
            if ($callerText === '') return;

            try {
                $reply = $this->classifyIntent($callerText);
            } catch (\Throwable $e) {
                fwrite(STDERR, "[!] classify failed: {$e->getMessage()}\n");
                $this->send($conn, ['command' => 'text', 'text' => 'Sorry, please try again.', 'isLast' => true]);
                return;
            }

            $intentKey = strtolower($reply);
            if (isset($this->cfg['agents'][$intentKey])) {        // one-word, known route
                [, $label] = $this->cfg['agents'][$intentKey];
                $this->send($conn, ['command' => 'text',
                    'text' => "Please wait, connecting you to $label.", 'isLast' => true]);

                $callId = $st['callId'];
                $st['patched'] = true;                            // guard immediately
                $this->state[$conn] = $st;

                $doPatch = function () use ($conn, $callId, $intentKey, $label) {
                    try {
                        $this->patchInAgent($callId, $intentKey);
                        echo "[*] patched $label into call $callId\n";
                    } catch (\Throwable $e) {
                        fwrite(STDERR, "[!] patch failed: {$e->getMessage()}\n");
                    }
                    $conn->close();
                };
                // Let the TTS play before we drop out; fall back to inline if no loop.
                if ($this->loop) { $this->loop->addTimer(2.5, $doPatch); } else { sleep(2); $doPatch(); }
            } else {                                              // not confident — clarify
                $this->send($conn, ['command' => 'text', 'text' => $reply, 'isLast' => true]);
            }
        }
    }

    public function onClose(ConnectionInterface $conn): void
    {
        $callId = $this->state[$conn]['callId'] ?? null;
        $this->state->detach($conn);
        echo "[-] session ended  callId=$callId\n";
    }

    public function onError(ConnectionInterface $conn, \Exception $e): void
    {
        fwrite(STDERR, "[!] ws error: {$e->getMessage()}\n");
        $conn->close();
    }

    private function send(ConnectionInterface $conn, array $payload): void
    {
        $raw = json_encode($payload);
        echo "  >> $raw\n";
        $conn->send($raw);
    }

    private function classifyIntent(string $callerText): string
    {
        $body = json_encode([
            'model'       => $this->cfg['llmModel'],
            'temperature' => 0,
            'max_tokens'  => 20,
            'messages'    => [
                ['role' => 'system', 'content' => $this->cfg['systemPrompt']],
                ['role' => 'user',   'content' => $callerText],
            ],
        ]);
        [$status, $resp] = $this->httpPost(
            "{$this->cfg['llmBaseUrl']}/chat/completions", $body,
            ['Authorization: Bearer ' . $this->cfg['llmApiKey'], 'Content-Type: application/json']);
        if ($status < 200 || $status >= 300) throw new \RuntimeException("LLM $status: $resp");
        $data = json_decode($resp, true);
        return trim($data['choices'][0]['message']['content'] ?? '');
    }

    private function patchInAgent(string $callId, string $intentKey): void
    {
        [$number, $label] = $this->cfg['agents'][$intentKey];
        $body = json_encode([
            'commands' => [[
                'command'  => 'dial',
                'callName' => 'agent_call',
                'from' => ['type' => 'PHONE', 'phone' => ['number' => $this->cfg['sinchNumber']]],
                'to'   => ['type' => 'PHONE', 'phone' => ['number' => $number]],
                'dialTimeoutDurationSeconds' => 20,
                'maxCallDurationSeconds' => 3600,
                'events' => [
                    'onAnswer' => [
                        ['command' => 'bridgeCall', 'bridgeName' => 'ivr-bridge'],
                        ['command' => 'messages', 'messagesName' => 'agent-intro',
                         'messages' => [['type' => 'SAY', 'say' => [
                             'format' => 'TEXT',
                             'text' => "Connecting you to a customer. Intent: $label.",
                             'voiceName' => 'Tiffany']]]],
                    ],
                    'onHangup' => [['command' => 'hangup', 'callName' => 'caller']],
                ],
            ]],
        ]);
        $auth = base64_encode("{$this->cfg['keyId']}:{$this->cfg['keySecret']}");
        [$status, $resp] = $this->httpPost(
            SINCH_BASE . "/projects/{$this->cfg['projectId']}/calls/$callId", $body,
            ['Authorization: Basic ' . $auth, 'Content-Type: application/json',
             "Idempotency-Key: $callId-$intentKey"], 'PATCH');
        if ($status !== 202) throw new \RuntimeException("PATCH $status: $resp");
    }

    private function httpPost(string $url, string $body, array $headers, string $method = 'POST'): array
    {
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_CUSTOMREQUEST  => $method,
            CURLOPT_POSTFIELDS     => $body,
            CURLOPT_HTTPHEADER     => $headers,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 10,
        ]);
        $resp = curl_exec($ch);
        if ($resp === false) { $err = curl_error($ch); curl_close($ch); throw new \RuntimeException($err); }
        $status = curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        curl_close($ch);
        return [$status, $resp];
    }
}

// ── Boot ─────────────────────────────────────────────────────────────────────
$relay = new IvrRelay($config);
$server = IoServer::factory(new HttpServer(new WsServer($relay)), $port, '0.0.0.0');
$relay->setLoop($server->loop);
echo "[*] AI IVR relay  model={$config['llmModel']}  listening on ws://0.0.0.0:$port\n";
$server->run();

File: 4.5-ai-ivr/scripts/relay-server.py
#!/usr/bin/env python3
"""
3.4.5 AI IVR — Voice Relay server that classifies caller intent with an LLM and
patches a human agent into the live call.

Flow per connection:
  1. Sinch sends {"command":"connect", "callId": "..."} -> we answer + greet.
  2. Sinch sends {"command":"text"/"prompt", "text":"<caller speech>"}.
  3. We ask the LLM to classify. One-word reply ("Sales"/"Support") = route;
     anything longer is spoken back as a clarifying question.
  4. On a route, we PATCH /v2/projects/{projectId}/calls/{callId} to dial the
     agent and bridge them into "ivr-bridge", then close the socket.

Requirements: pip install -r ../requirements.txt   (websockets, requests)
Run:    python relay-server.py
Expose: ngrok http 8765
"""

import asyncio
import json
import os

import requests
import websockets

# ── Load the tutorial-folder .env (../.env relative to this scripts/ folder) ──
_env_path = os.path.join(os.path.dirname(__file__), "..", ".env")
if os.path.exists(_env_path):
    with open(_env_path) as _f:
        for _line in _f:
            _line = _line.strip()
            if not _line or _line.startswith("#") or "=" not in _line:
                continue
            _k, _v = _line.split("=", 1)
            os.environ.setdefault(_k.strip(), _v.strip().strip('"').strip("'"))

# ── Configuration ────────────────────────────────────────────────────────────
def _require(name: str) -> str:
    val = os.environ.get(name)
    if not val:
        raise SystemExit(f"ERROR: {name} is not set in the environment / .env")
    return val

PROJECT_ID  = _require("PROJECT_ID")
KEY_ID      = _require("KEY_ID")
KEY_SECRET  = _require("KEY_SECRET")
SINCH_NUMBER   = _require("SINCH_NUMBER")
SALES_NUMBER   = _require("SALES_NUMBER")
SUPPORT_NUMBER = _require("SUPPORT_NUMBER")
LLM_BASE_URL = os.environ.get("LLM_BASE_URL", "https://api.openai.com/v1").rstrip("/")
LLM_API_KEY  = _require("LLM_API_KEY")
LLM_MODEL    = os.environ.get("LLM_MODEL", "gpt-4o-mini")
PORT         = int(os.environ.get("PORT", "8765"))

SINCH_BASE = "https://voice.api.sinch.com/v2"
GREETING   = "Hello, this is the call centre. How can I help you?"

AGENTS = {"sales": (SALES_NUMBER, "Sales"), "support": (SUPPORT_NUMBER, "Support")}

_prompt_path = os.path.join(os.path.dirname(__file__), "..", "system_prompt.md")
with open(_prompt_path) as _f:
    SYSTEM_PROMPT = _f.read().strip()


# ── LLM intent classification (blocking; called via asyncio.to_thread) ───────
def classify_intent(caller_text: str) -> str:
    """Return the model's reply: one word ('Sales'/'Support') or a clarification."""
    resp = requests.post(
        f"{LLM_BASE_URL}/chat/completions",
        headers={"Authorization": f"Bearer {LLM_API_KEY}",
                 "Content-Type": "application/json"},
        json={
            "model": LLM_MODEL,
            "temperature": 0,
            "max_tokens": 20,
            "messages": [
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": caller_text},
            ],
        },
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()["choices"][0]["message"]["content"].strip()


# ── PATCH the live call to bridge in a human agent (blocking) ────────────────
def patch_in_agent(call_id: str, intent_key: str) -> None:
    number, label = AGENTS[intent_key]
    body = {
        "commands": [
            {
                "command": "dial",
                "callName": "agent_call",
                "from": {"type": "PHONE", "phone": {"number": SINCH_NUMBER}},
                "to":   {"type": "PHONE", "phone": {"number": number}},
                "dialTimeoutDurationSeconds": 20,
                "maxCallDurationSeconds": 3600,
                "events": {
                    "onAnswer": [
                        {"command": "bridgeCall", "bridgeName": "ivr-bridge"},
                        {"command": "messages", "messagesName": "agent-intro",
                         "messages": [{"type": "SAY", "say": {
                             "format": "TEXT",
                             "text": f"Connecting you to a customer. Intent: {label}.",
                             "voiceName": "Tiffany"}}]},
                    ],
                    "onHangup": [{"command": "hangup", "callName": "caller"}],
                },
            }
        ]
    }
    resp = requests.patch(
        f"{SINCH_BASE}/projects/{PROJECT_ID}/calls/{call_id}",
        auth=(KEY_ID, KEY_SECRET),
        headers={"Content-Type": "application/json",
                 "Idempotency-Key": f"{call_id}-{intent_key}"},
        json=body,
        timeout=10,
    )
    if resp.status_code != 202:
        raise RuntimeError(f"PATCH returned {resp.status_code}: {resp.text}")


# ── WebSocket session ────────────────────────────────────────────────────────
async def send(ws, payload: dict) -> None:
    raw = json.dumps(payload)
    print(f"  >> {raw}")
    await ws.send(raw)


async def handle_connection(ws):
    call_id = None
    patched = False
    print(f"[+] connected: {ws.remote_address}")
    try:
        async for raw in ws:
            print(f"  << {raw}")
            try:
                msg = json.loads(raw)
            except json.JSONDecodeError:
                continue
            command = msg.get("command")

            if command == "connect":
                call_id = msg.get("callId")
                await send(ws, {"command": "answer"})
                await send(ws, {"command": "text", "text": GREETING, "isLast": True})

            elif command in ("text", "prompt") and call_id and not patched:
                caller_text = (msg.get("text") or "").strip()
                if not caller_text:
                    continue
                try:
                    reply = await asyncio.to_thread(classify_intent, caller_text)
                except Exception as exc:
                    print(f"[!] classify failed: {exc}")
                    await send(ws, {"command": "text",
                                    "text": "Sorry, please try again.", "isLast": True})
                    continue

                intent_key = reply.lower()
                if intent_key in AGENTS:  # one-word, known route
                    _, label = AGENTS[intent_key]
                    await send(ws, {"command": "text",
                                    "text": f"Please wait, connecting you to {label}.",
                                    "isLast": True})
                    await asyncio.sleep(2.5)  # let the TTS play before we drop out
                    try:
                        await asyncio.to_thread(patch_in_agent, call_id, intent_key)
                        patched = True
                        print(f"[*] patched {label} into call {call_id}")
                    except Exception as exc:
                        print(f"[!] patch failed: {exc}")
                    await ws.close()
                    return
                else:  # not confident — speak the clarification back
                    await send(ws, {"command": "text", "text": reply, "isLast": True})

    except websockets.exceptions.ConnectionClosed as exc:
        print(f"[-] closed ({exc.code}): {exc.reason}")
    finally:
        print(f"[-] session ended  callId={call_id}")


async def main():
    print(f"[*] AI IVR relay  model={LLM_MODEL}  listening on ws://0.0.0.0:{PORT}")
    async with websockets.serve(handle_connection, "0.0.0.0", PORT):
        await asyncio.Future()


if __name__ == "__main__":
    asyncio.run(main())
